[inductor] weight prepack for _convolution_transpose_pointwise

chunyuan-w opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90267
* __->__ #90266
* #90265
* #90264


This PR implements weight prepack for `_convolution_transpose_pointwise`, similar to https://github.com/pytorch/pytorch/pull/88988.

In addition, we add a kernel for Meta tensor input to reduce the compilation time.

cc @VitalyFedyunin @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

chunyuan-w committed:
[inductor] weight prepack for _convolution_transpose_pointwise

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] weight prepack for _convolution_transpose_pointwise"




cc VitalyFedyunin jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] weight prepack for _convolution_transpose_pointwise"




cc VitalyFedyunin jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] weight prepack for _convolution_transpose_pointwise"




cc VitalyFedyunin jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] weight prepack for _convolution_transpose_pointwise"




cc VitalyFedyunin jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] weight prepack for _convolution_transpose_pointwise"




cc VitalyFedyunin jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -643,22 +647,42 @@ Tensor _mkldnn_convolution_transpose(
+    // weight is the prepacked mkldnn tensor
+    w = itensor_from_tensor(weight);
+  } else {
+    // aten tensor with the shape of the prepacked tensor
+    at::Tensor origin_weight_t;
+    if (groups > 1) {
+      origin_weight_t = weight.transpose(0, 1).reshape(weight_IOHW_sizes);
+    } else {
+      origin_weight_t = weight.transpose(0, 1);
+    }
+    w = itensor_from_tensor(origin_weight_t);
+    w.transpose_(0, 1);
```
Do we have a case that we want to handle this "aten tensor with the shape of the prepacked tensor", or we always model prepacked weight with "MKLDNNTensor"?

chunyuan-w replied:
When compiling the graph, each node is run with FakeTensor inputs and in this case, the weight tensor is an aten tensor with the shape of the prepacked tensor.

chunyuan-w replied:
For Meta tensor: added an implementation `mkldnn_convolution_transpose_pointwise_meta`.

For MKLDNN and CPU tensor: supported the computation for both of them in the original implementation: `mkldnn_convolution_transpose_pointwise`.

chunyuan-w committed:
Update on "[inductor] weight prepack for _convolution_transpose_pointwise"




cc VitalyFedyunin jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] weight prepack for _convolution_transpose_pointwise"



This PR implements weight prepack for `_convolution_transpose_pointwise`, similar to https://github.com/pytorch/pytorch/pull/88988.


Different from Conv2d, once ConvTranspose weight has been prepacked, the size changes:
- Original weight size: `[i, o, ...]`
- Prepacked size:

  - Groups > 1:  `[g*o, i/g, ...]`
  - Groups == 1: `[o, i, ...]`

The  `_convolution_transpose_pointwise` kernel handles the below two situations:

- During compilation, when running the node of the FX graph, the kernel gets a public weight tensor with prepacked size. The kernel will convert the weight back to the original weight size for computation.
- During execution, it gets a mkldnn tensor and will directly use it for computation.

cc VitalyFedyunin jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] weight prepack for _convolution_transpose_pointwise"



This PR implements weight prepack for `_convolution_transpose_pointwise`, similar to https://github.com/pytorch/pytorch/pull/88988.


Different from Conv2d, once ConvTranspose weight has been prepacked, the size changes:
- Original weight size: `[i, o, ...]`
- Prepacked size:

  - Groups > 1:  `[g*o, i/g, ...]`
  - Groups == 1: `[o, i, ...]`

The  `_convolution_transpose_pointwise` kernel handles the below two situations:

- During compilation, when running the node of the FX graph, the kernel gets a public weight tensor with prepacked size. The kernel will convert the weight back to the original weight size for computation.
- During execution, it gets a mkldnn tensor and will directly use it for computation.

cc VitalyFedyunin jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] weight prepack for _convolution_transpose_pointwise"



This PR implements weight prepack for `_convolution_transpose_pointwise`, similar to https://github.com/pytorch/pytorch/pull/88988.


Different from Conv2d, once ConvTranspose weight has been prepacked, the size changes:
- Original weight size: `[i, o, ...]`
- Prepacked size:

  - Groups > 1:  `[g*o, i/g, ...]`
  - Groups == 1: `[o, i, ...]`

The  `_convolution_transpose_pointwise` kernel handles the below two situations:

- During compilation, when running the node of the FX graph, the kernel gets a public weight tensor with prepacked size. The kernel will convert the weight back to the original weight size for computation.
- During execution, it gets a mkldnn tensor and will directly use it for computation.

cc VitalyFedyunin jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -643,22 +648,33 @@ Tensor _mkldnn_convolution_transpose(
+
+  const ideep::tensor x = itensor_from_tensor(input);
+
+  ideep::tensor w = itensor_from_tensor(weight);;
```
```suggestion
  ideep::tensor w = itensor_from_tensor(weight);
```

chunyuan-w replied:
Fixed the typo.

jgong5 commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -709,6 +725,8 @@ Tensor mkldnn_convolution_transpose_pointwise(
     c10::optional<c10::string_view> algorithm) {
   c10::impl::ExcludeDispatchKeyGuard edkg(c10::autograd_dispatch_keyset);
+  bool use_channels_last =
+      weight_t.is_mkldnn() || mkldnn_conv_use_channels_last(input_t, weight_t);
```
The logic is ok. Not related to this PR, but does it make more sense that `mkldnn_conv_use_channels_last` should prefer channels last if input is a strided tensor and weight is mkldnn? @XiaobingSuper

jgong5 commented on torch/_inductor/ir.py:
```diff
@@ -3349,6 +3349,53 @@ def get_template_tiling(self):
 
 
+# Port from aten/src/ATen/native/ConvUtils.h: _conv_input_size
+def _conv_input_size(
+    output_size, weight_size, padding, output_padding, stride, dilation, groups
+):
```
How about we keep this function local to `_prepare_convolution_fusion_create` first? It is only used by this function.

chunyuan-w replied:
Made this function local to `_prepare_convolution_fusion_create`.

jgong5 commented on torch/_inductor/ir.py:
```diff
@@ -3349,6 +3349,53 @@ def get_template_tiling(self):
+#   Groups > 1:  [g*o, i/g, ...]
+#   Groups == 1: [o, i, ...]
+# Returns original weight size in [i, o, ...]
+def _original_deconv_weight_size(
+    prepacked_weight,
+    groups,
+):
```
ditto.

chunyuan-w replied:
Fixed.

chunyuan-w committed:
Update on "[inductor] weight prepack for _convolution_transpose_pointwise"



This PR implements weight prepack for `_convolution_transpose_pointwise`, similar to https://github.com/pytorch/pytorch/pull/88988.

In addition, we add a kernel for Meta tensor input to reduce the compilation time.

cc VitalyFedyunin jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] weight prepack for _convolution_transpose_pointwise"



This PR implements weight prepack for `_convolution_transpose_pointwise`, similar to https://github.com/pytorch/pytorch/pull/88988.

In addition, we add a kernel for Meta tensor input to reduce the compilation time.

cc VitalyFedyunin jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] weight prepack for _convolution_transpose_pointwise"



This PR implements weight prepack for `_convolution_transpose_pointwise`, similar to https://github.com/pytorch/pytorch/pull/88988.

In addition, we add a kernel for Meta tensor input to reduce the compilation time.

cc VitalyFedyunin jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] weight prepack for _convolution_transpose_pointwise"



This PR implements weight prepack for `_convolution_transpose_pointwise`, similar to https://github.com/pytorch/pytorch/pull/88988.

In addition, we add a kernel for Meta tensor input to reduce the compilation time.

cc VitalyFedyunin jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] weight prepack for _convolution_transpose_pointwise"



This PR implements weight prepack for `_convolution_transpose_pointwise`, similar to https://github.com/pytorch/pytorch/pull/88988.

In addition, we add a kernel for Meta tensor input to reduce the compilation time.

cc VitalyFedyunin jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] weight prepack for _convolution_transpose_pointwise"



This PR implements weight prepack for `_convolution_transpose_pointwise`, similar to https://github.com/pytorch/pytorch/pull/88988.

In addition, we add a kernel for Meta tensor input to reduce the compilation time.

cc VitalyFedyunin jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] weight prepack for _convolution_transpose_pointwise"



This PR implements weight prepack for `_convolution_transpose_pointwise`, similar to https://github.com/pytorch/pytorch/pull/88988.

In addition, we add a kernel for Meta tensor input to reduce the compilation time.

cc VitalyFedyunin jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] weight prepack for _convolution_transpose_pointwise"



This PR implements weight prepack for `_convolution_transpose_pointwise`, similar to https://github.com/pytorch/pytorch/pull/88988.

In addition, we add a kernel for Meta tensor input to reduce the compilation time.

cc VitalyFedyunin jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "[inductor] weight prepack for _convolution_transpose_pointwise"



This PR implements weight prepack for `_convolution_transpose_pointwise`, similar to https://github.com/pytorch/pytorch/pull/88988.

In addition, we add a kernel for Meta tensor input to reduce the compilation time.

cc VitalyFedyunin jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w commented:
Re-opened in https://github.com/pytorch/pytorch/pull/91955
